### PR TITLE
Filter `null` Ad Slot Refs

### PR DESCRIPTION
**Pull request by JamieB-gu**
The way the ad slot refs were created meant that they were stored as possibly `null` values, which complicated the code that needed to use them. This change makes use of a callback function to set the refs, meaning they can never be `null`. This has a few advantages:

- The code using the slots doesn't have to check to see if they're `null`
- The setup code becomes simpler as we no longer need to use `createRef` in a loop
- The type of the `adSlots` ref is simpler
- We can now check to make sure all the ad slot refs have been created correctly before inserting ads or setting up the ad update code

### CODE REVIEW

File: dotcom-rendering/src/components/AdPortals.importable.tsx [hunk 1]
```diff
@@ -158,14 +103,52 @@ export const AdPortals = () => {
...
+						adPositions.current = newAdPositions;
+					}
+				}
+			});
+
+			resizeObserver.observe(document.body);
 		}
+
+		return () => resizeObserver?.disconnect();
 	}, [adPlaceholders]);
```
Comment (abeddow91) [inline dotcom-rendering/src/components/AdPortals.importable.tsx:152]: If we are dependent on adPlaceholders here, does that mean we're destroying and  recreating the resize observer when the adPlaceholders array changes? I guess this would only be once so maybe this is not much of a concern

File: dotcom-rendering/src/components/AdPortals.importable.tsx [hunk 2]
```diff
@@ -158,14 +103,52 @@ export const AdPortals = () => {
...
+
+		if (
+			adPlaceholders.length !== 0 &&
+			adPlaceholders.length === adSlots.current.length
+		) {
+			adPositions.current = calculateAdPositions(adSlots.current);
+			bodyHeight.current = document.body.clientHeight;
+			void getCommercialClient().insertAdverts(adPositions.current);
+
+			resizeObserver = new ResizeObserver((entries) => {
```
Comment (abeddow91) [inline dotcom-rendering/src/components/AdPortals.importable.tsx:131]: If we are dependent on adPlaceholders here, does that mean we're destroying and recreating the resize observer when the adPlaceholders array changes? I guess this would only be once so maybe this is not much of a concern

File: dotcom-rendering/src/components/AdPortals.importable.tsx [hunk 3]
```diff
@@ -177,7 +160,11 @@ export const AdPortals = () => {
 						isHidden={false}
 						isSquare={index === 0}
 						index={index}
-						ref={adSlots.current[index]}
+						ref={(node) => {
+							if (node !== null) {
+								adSlots.current = [...adSlots.current, node];
+							}
+						}}
```
Comment (abeddow91) [inline dotcom-rendering/src/components/AdPortals.importable.tsx:167]: ❤️

**Review by abeddow91 — APPROVED** [review]
This looks great! I've one comment around moving the resize observer set up into the adplaceholder dependent useeffect but its non blocking

File: dotcom-rendering/src/components/AdPortals.importable.tsx (hunk 2 above)
Comment (JamieB-gu) [reply dotcom-rendering/src/components/AdPortals.importable.tsx:131]: I'd expect this `useEffect` to run twice over the lifetime of the component:

1. First render: this `useEffect` runs but the list of ad placeholders is empty (length is 0), so the resize observer isn't created.
2. Subsequent render caused by the `adPlaceholders` state changing: the list of ad placeholders is now populated so this useEffect runs again and this time the resize observer is created.

`adPlaceholders` should never change from this point, so this `useEffect` shouldn't run again.

In conclusion, the resize observer should only ever be created once. However, even if this `useEffect` does end up running more than expected, the clean-up code means that the resize observer should be safely torn down and re-created.
